pkg/lib/packagemanager.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2025 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+
+import { PackageManager, Enum, UnsupportedError, NotFoundError } from './_internal/packagemanager-abstract';
+import { Dnf5DaemonManager } from './_internal/dnf5daemon';
+import { PackageKitManager } from './_internal/packagekit';
+
+function debug(...args: unknown[]) {
+    if (window.debugging == 'all' || window.debugging?.includes('packagemanager'))
+        console.debug('packagemanager', ...args);
+}
+
+async function is_immutable_os() {
+    try {
+        const options = await cockpit.spawn(["findmnt", "-T", "/usr", "-n", "-o", "VFS-OPTIONS"]);
+        return options.split(",").indexOf("ro") >= 0;
+    } catch (err) {
+        console.warn("Unable to detect immutable OS", err);
+        return false;
+    }
+}
+
+async function detect_dnf5daemon() {
+    try {
+        const client = cockpit.dbus("org.rpm.dnf.v0", { superuser: "try" });
+        await client.call("/org/rpm/dnf/v0", "org.freedesktop.DBus.Peer", "Ping", []);
+        return true;
+    } catch (err) {
+        console.warn("dnf5daemon not supported", err);
+        return false;
+    }
+}
+
+async function detect_packagekit() {
+    try {
+        const client = cockpit.dbus("org.freedesktop.PackageKit", { superuser: "try" });
+        await client.call("/org/freedesktop/PackageKit", "org.freedesktop.DBus.Properties",
+                          "Get", ["org.freedesktop.PackageKit", "VersionMajor"]);
+        return true;
+    } catch (err) {
+        console.warn("PackageKit not supported", err);
+        return false;
+    }
+}
+
+// Cache result for a session
+export async function getPackageManager(): Promise<PackageManager> {
+    const [unsupported, has_dnf5daemon, has_packagekit] = await Promise.all([is_immutable_os(), detect_dnf5daemon(), detect_packagekit()]);
+
+    if (unsupported)
+        throw new UnsupportedError("Unsupported package manager");
```

Comment:
Can you please make this more direct, with `isImmutable` and mentioning this in the error message?